File: token_types.c
#include <string.h>
#include "token_types.h"

const char *token_type_name(TokenType type) {
    switch (type) {
        case KW_BOOL: return "KW_BOOL";
        case KW_INT: return "KW_INT";
        case KW_FLOAT: return "KW_FLOAT";
        case KW_CHAR: return "KW_CHAR";
        case KW_DOUBLE: return "KW_DOUBLE";
        case KW_VOID: return "KW_VOID";
        case KW_TRUE: return "KW_TRUE";
        case KW_FALSE: return "KW_FALSE";
        case KW_IF: return "KW_IF";
        case KW_ELSE: return "KW_ELSE";
        case KW_WHILE: return "KW_WHILE";
        case KW_FOR: return "KW_FOR";
        case KW_RETURN: return "KW_RETURN";
        case KW_BREAK: return "KW_BREAK";
        case KW_CONTINUE: return "KW_CONTINUE";
        case KW_PRINT: return "KW_PRINT";
        case KW_READ: return "KW_READ";
        case ID: return "ID";
        case NUM_INT: return "NUM_INT";
        case NUM_FLOAT: return "NUM_FLOAT";
        case STRING: return "STRING";
        case CHAR_LITERAL: return "CHAR_LITERAL";
        case PLUS: return "PLUS";
        case MINUS: return "MINUS";
        case STAR: return "STAR";
        case SLASH: return "SLASH";
        case PERCENT: return "PERCENT";
        case ASSIGN: return "ASSIGN";
        case EQ: return "EQ";
        case NEQ: return "NEQ";
        case LT: return "LT";
        case LE: return "LE";
        case GT: return "GT";
        case GE: return "GE";
        case AND: return "AND";
        case OR: return "OR";
        case NOT: return "NOT";
        case LPAREN: return "LPAREN";
        case RPAREN: return "RPAREN";
        case LBRACE: return "LBRACE";
        case RBRACE: return "RBRACE";
        case LBRACKET: return "LBRACKET";
        case RBRACKET: return "RBRACKET";
        case SEMI: return "SEMI";
        case COMMA: return "COMMA";
        case DOT: return "DOT";
        case TOK_EOF: return "EOF";
        case TOK_ERROR: return "ERROR";
    }
    return "UNKNOWN";
}

typedef struct {
    const char *word;
    TokenType type;
} ReservedWord;

/* Equivalente ao dicionario RESERVED_WORDS do Python */
static const ReservedWord RESERVED_WORDS[] = {
    {"bool", KW_BOOL},
    {"int", KW_INT},
    {"float", KW_FLOAT},
    {"char", KW_CHAR},
    {"double", KW_DOUBLE},
    {"void", KW_VOID},
    {"true", KW_TRUE},
    {"false", KW_FALSE},
    {"if", KW_IF},
    {"else", KW_ELSE},
    {"while", KW_WHILE},
    {"for", KW_FOR},
    {"return", KW_RETURN},
    {"break", KW_BREAK},
    {"continue", KW_CONTINUE},
    {"print", KW_PRINT},
    {"read", KW_READ},
};

#define NUM_RESERVED_WORDS (sizeof(RESERVED_WORDS) / sizeof(RESERVED_WORDS[0]))

int is_reserved_word(const char *word, TokenType *out_type) {
    for (size_t i = 0; i < NUM_RESERVED_WORDS; i++) {
        if (strcmp(RESERVED_WORDS[i].word, word) == 0) {
            *out_type = RESERVED_WORDS[i].type;
            return 1;
        }
    }
    return 0;
}

TokenType lookup_reserved_word(const char *word) {
    TokenType t;
    if (is_reserved_word(word, &t)) {
        return t;
    }
    return ID;
}

File: ProjetoMiniC/Python/src/lexer/token_types.py
"""
token_types.py
==============

Define os tipos de token (`TokenType`) e o dicionário de palavras
reservadas (`RESERVED_WORDS`) da linguagem MiniC.

Este módulo é a base do pacote: não depende de nenhum outro módulo
interno, apenas da biblioteca padrão (`enum`, `typing`).
"""

from enum import Enum, auto
from typing import Dict


class TokenType(Enum):
    """Enumeração de todos os tipos de token reconhecidos pelo scanner."""

    # ------------------------------------------------------------------
    # A enumeração centraliza todas as categorias léxicas da linguagem.
    # Cada valor representa um tipo de token que o scanner pode produzir.
    # Isso facilita a análise sintática e a geração de diagnósticos.
    # ------------------------------------------------------------------

    # Palavras reservadas
    KW_BOOL = auto()
    KW_INT = auto()
    KW_FLOAT = auto()
    KW_CHAR = auto()
    KW_DOUBLE = auto()
    KW_VOID = auto()
    KW_TRUE = auto()
    KW_FALSE = auto()
    KW_IF = auto()
    KW_ELSE = auto()
    KW_WHILE = auto()
    KW_FOR = auto()
    KW_RETURN = auto()
    KW_BREAK = auto()
    KW_CONTINUE = auto()
    KW_PRINT = auto()
    KW_READ = auto()

    # Identificador
    ID = auto()

    # Literais
    NUM_INT = auto()
    NUM_FLOAT = auto()
    STRING = auto()
    CHAR_LITERAL = auto()

    # Operadores aritméticos
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()

    # Atribuição e relacionais
    ASSIGN = auto()
    EQ = auto()
    NEQ = auto()
    LT = auto()
    LE = auto()
    GT = auto()
    GE = auto()

    # Operadores lógicos
    AND = auto()   # &&
    OR = auto()    # ||
    NOT = auto()   # !

    # Delimitadores
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    SEMI = auto()
    COMMA = auto()
    DOT = auto()

    # Controle
    EOF = auto()
    ERROR = auto()   # token inválido, mantido na listagem para diagnóstico


# ------------------------------------------------------------------
# Dicionário que mapeia as palavras reservadas da linguagem MiniC para
# seus respectivos tipos. Quando o scanner encontra uma sequência como
# "if" ou "return", ele reconhece imediatamente o token correto.
# ------------------------------------------------------------------
RESERVED_WORDS: Dict[str, TokenType] = {
    "bool": TokenType.KW_BOOL,
    "int": TokenType.KW_INT,
    "float": TokenType.KW_FLOAT,
    "char": TokenType.KW_CHAR,
    "double": TokenType.KW_DOUBLE,
    "void": TokenType.KW_VOID,
    "true": TokenType.KW_TRUE,
    "false": TokenType.KW_FALSE,
    "if": TokenType.KW_IF,
    "else": TokenType.KW_ELSE,
    "while": TokenType.KW_WHILE,
    "for": TokenType.KW_FOR,
    "return": TokenType.KW_RETURN,
    "break": TokenType.KW_BREAK,
    "continue": TokenType.KW_CONTINUE,
    "print": TokenType.KW_PRINT,
    "read": TokenType.KW_READ,
}
